Add support for ingestion of out-of-order native histogram samples

carrieedwards opened the pull request:
This PR adds initial support for out-of-order ingestion of native histograms. It also includes unit tests for ingestion, compaction, and querying of OOO native histogram samples and chunks. 

Relevant issue: [#11220](https://github.com/prometheus/prometheus/issues/11220)

zenador committed:
Start to modify write path to support OOO for native histograms

Signed-off-by: Jeanette Tan <jeanette.tan@grafana.com>

Add unit test for observing out of bound histograms

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

TestOOOWALWrite: make generic and add integer histogram case

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Make TestOOOWalWrite pass with integer native histograms

First version of the code. WIP.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Fix lint error

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Do not use 0 reference to indicate no mmap reference

Now we can use nil, since we're returning a slice.
Fixes broken TestOOOCompactionWithDisabledWriteLog

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Fix handling case of recording zero chunk written to disk.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Implement float histograms into WBL at append

This is the dumb copy paste implementation to get tests to pass.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Test_ChunkQuerier_OOOQuery: add integer and float histograms

Test with integer and float histograms
Counter only.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Have a single OOOChunk.ToEncodedChunk function

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Make simple query work

OOOHeadChunk yields a single encoded chunk.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Add TODO

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Update tsdb/ooo_head.go

Use common extractor function for testcases

Probably replicating some code here, this can't be the first place for this.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Fix updating appender when opening new chunk due to counterreset

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Fix writing chunk times to disk when multiple chunks from ooo head

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Allow reading multiple chunks from the tsdb ooo head

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Fix linter

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Update docstring wording

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

Follow up merge from main

Adopt #12352 improvement.

Signed-off-by: György Krajcsovits <gyorgy.krajcsovits@grafana.com>

zenador committed:
Update OOO tests

Signed-off-by: Jeanette Tan <jeanette.tan@grafana.com>

fionaliao committed:
Add OOO histogram cases to head_test.go

- Add histogram test cases
- Add TestOOOAppendWithNoSeries test
- Extend TestOOOMmapReplay with histograms

carrieedwards committed:
Add OOO histogram cases to db_test.go

- Extend TestOOODisabled with histograms
- Add histogram test cases for Test_Querier_OOOQuery
- Add histogram test cases for TestOOOCompaction
- Add histogram test cases for TestOOOCompactionWithNormalCompaction
- Add histogram test cases for TestOOOCompactionWithDisabledWriteLog
- Add histogram test cases for TestOOOCompactionFailure

fionaliao committed:
Process histogram samples in the WBL

Also added a test case.

Signed-off-by: Fiona Liao <fiona.y.liao@gmail.com>

carrieedwards committed:
Fix handling of encoding for histograms in mergedOOOChunks and boundedIterable

- Fix encoding
- Handle chunk valueType in boundedIterator
- Reduce code duplication
- Updated boundedChunk test with native histogram testcases
- Fix expected samples for boundedChunk test
- Use AtT() in boundedChunk seek

fionaliao committed:
Add more OOO histogram cases to db_test.go

Extend TestDiskFillingUpAfterDisablingOOO

Extend TestPanicOnApplyConfig

Extend TestNoGapAfterRestartWithOOO and TestWblReplayAfterOOODisableAndRestart

Extend TestOutOfOrderRuntimeConfig

Extend TestOOOMmapCorruption

Extend TestWBLAndMmapReplay

Extend TestOOOAppendAndQuery

Extend TestOOOQueryAfterRestartWithSnapshotAndRemovedWBL

carrieedwards committed:
Use test helper functions to avoid code duplication

carrieedwards committed:
Add histogram test cases to OOO head inserts and reads

Add histogram test cases to OOO head insertion tests

Fix linting

Add native histogram testcases for OOO head reads

carrieedwards committed:
Refactor TSDB test helper structs/functions into separate file

carrieedwards committed:
Fix tests after rebasing main

Fix OOO WAL write test

Fix WBL replay test

Fix test

fionaliao committed:
Add counter reset test for mmapped OOO chunks (#544)

* Add counter reset test for mmapped OOO chunks

Fixed a couple of bugs that were discovered while writing the test:
- mint for ooo chunk needed to be set when a new chunk was created
- first chunk was using the wrong previous appender

Signed-off-by: Fiona Liao <fiona.y.liao@gmail.com>

* nolint:staticcheck

---------

Signed-off-by: Fiona Liao <fiona.y.liao@gmail.com>

carrieedwards committed:
Add tests for gauge histograms

Refactor sample comparison util and use in head_test.go

Linting

fionaliao committed:
Improve requireEqualSamples (#565)

- Allow counter reset header to be compared
- Add timestamp value check
- Add timestamp to error message

Signed-off-by: Fiona Liao <fiona.y.liao@gmail.com>

fionaliao committed:
Add test for counter resets when compacting ooo native histograms (#574)

* Add test for ooo histogram compaction

Signed-off-by: Fiona Liao <fiona.y.liao@gmail.com>

carrieedwards committed:
Add engine tests for OOO native histograms with counter resets (#555)

* Add engine tests for OOO native histograms with counter resets

* Expand test cases and refactor

* Add in-order and OOO native histogram tests for increase function

* Add in-order and OOO native histogram tests for resets function

* Clean up tests

* Fix linting

* Test refactor and adding comments

* Test cleanup

* Fix lint errors

* Fix case where in-order samples were ingested instead

---------

Co-authored-by: Fiona Liao <fiona.y.liao@gmail.com>

fionaliao committed:
Update OOO counter reset test after validation changes (#581)

carrieedwards committed:
Fix linting

carrieedwards committed:
Add tests for ingesting and querying OOO Native Histograms with counter resets

carrieedwards committed:
More linting

carrieedwards committed:
Lint

carrieedwards committed:
Add flag for enabling OOO native histogram ingestion

fionaliao committed:
Set EnableOOONativeHistograms to true if flag supplied (#582)

fionaliao committed:
Add multiple encoding test (#579)

fionaliao committed:
Remove deleted flag

fionaliao commented on tsdb/head_read.go:
```diff
@@ -772,3 +766,45 @@ func makeStopIterator(c chunkenc.Chunk, it chunkenc.Iterator, stopAfter int) chu
 	}
 }
+
+func GetBytes(encoding chunkenc.Encoding, it chunkenc.Iterator) []byte {
```
Not used anymore, is removed in https://github.com/grafana/mimir-prometheus/pull/588

fionaliao committed:
Add OOO histogram-specific metrics

jesusvazquez commented on tsdb/head.go:
```diff
@@ -457,10 +462,10 @@ func newHeadMetrics(h *Head, r prometheus.Registerer) *headMetrics {
 			Help: "Total number of appended out of order samples.",
-		}),
+		}, []string{"type"}),
 		outOfBoundSamples: prometheus.NewCounterVec(prometheus.CounterOpts{
```
If we make `outOfOrderSamplesAppended` a countervec i think `outOfBoundSamples` should be a countervec too.

fionaliao replied:
`outOfBoundSamples` is already a countervec. Previously it only tracked oob floats though, this PR starts tracking oob histograms as well

jesusvazquez commented on tsdb/head_append.go:
```diff
@@ -581,10 +629,16 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
 
-	if t < a.minValidTime {
+	// For OOO inserts, this restriction is irrelevant and will be checked later once we confirm the histogram sample is an in-order append.
+	// If OOO inserts are disabled, we may as well as check this as early as we can and avoid more work.
```
I like this comment but I'd rephrase it to say something along the lines of "fail fast if ooo is disabled"

jesusvazquez commented on tsdb/head_append.go:
```diff
@@ -581,10 +629,16 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
 		a.head.metrics.outOfBoundSamples.WithLabelValues(sampleMetricTypeHistogram).Inc()
 		return 0, storage.ErrOutOfBounds
 	}
+	// If OOO is enabled, but OOO native histogram ingestion is disabled
```
Like above "also fail fast if ..."

fionaliao replied:
Updated comment in https://github.com/grafana/mimir-prometheus/pull/542/commits/ccbf860c6a9a78e4389e3ee4aea7ba8f8ff6f093

jesusvazquez commented on tsdb/head_append.go:
```diff
@@ -633,15 +687,25 @@ func (a *headAppender) AppendHistogram(ref storage.SeriesRef, lset labels.Labels
-		if err := s.appendableHistogram(t, h); err != nil {
-			s.Unlock()
-			if errors.Is(err, storage.ErrOutOfOrderSample) {
+		// TODO(codesome): If we definitely know at this point that the sample is ooo, then optimise
```
Does this todo still apply here?

fionaliao replied:
Yes - it's not been optimised yet. Currently, OOO samples are always appended to the WAL.

jesusvazquez commented on tsdb/head_append.go:
```diff
@@ -1463,33 +1713,41 @@ func (s *memSeries) cutNewHeadChunk(mint int64, e chunkenc.Encoding, chunkRange
+		handleChunkWriteError(err)
+		return nil
+	}
+	chunkRefs := make([]chunks.ChunkDiskMapperRef, 0, 1)
```
Could the list of chunk refs also be len(chks)

fionaliao replied:
Yep, updated in https://github.com/grafana/mimir-prometheus/pull/542/commits/ccbf860c6a9a78e4389e3ee4aea7ba8f8ff6f093

jesusvazquez reviewed (commented):
I've done a first pass to the ingest part. Tomorrow I'll do first pass to reads. 

Overall looks great nice work @fionaliao and @carrieedwards 

There is some code duplication, specially in the Commit() bits but I think its an issue with how the code was before, its hard to abstract it so we have less duplication at this point. It would be a big effort so we have to live with it.

fionaliao committed:
Clean up OOO native histograms test code (#588)

* Remove unused GetBytes() function

This is unnecessary after the ChunkOrIterable() refactoring (as oooMergedChunk doesn't need to implement the chunk interface anymore).

* Move err to last parameter of appendFunc

To make it more consistent with Go conventions.

* Move expOOOMmappedChunks defintion into test

Only used by TestOOOHistogramCounterResetHeaders so moving it within that test.

* Extract sample reading to a util function

fionaliao committed:
Address Jesus comments

krajorama commented on util/teststorage/storage.go:
```diff
@@ -50,6 +50,25 @@ func New(t testutil.T) *TestStorage {
 	return &TestStorage{DB: db, exemplarStorage: es, dir: dir}
 }
 
+func NewTestStorageWithOpts(t testutil.T, opts *tsdb.Options) *TestStorage {
```
Seems like the function `New` above could just use this if we didn't set OutOfOrderCapMax here?

carrieedwards committed:
Fix for float sample comparison in testutil
